### PR TITLE
Add NumPy style warning when casting complex to float

### PR DESCRIPTION
**Pull request by carlthome**
This is a real numeric gotcha, IMO.

I just used `tf.losses.mean_squared_error` with complex tensors and was pleasantly surprised when everything ran smoothly. However, it turns out that the function calls `tf.to_float` on its input tensors, which will just remove the imaginary part and cast the real part to float. I should preferably have been warned about this (like in NumPy) instead of reading through the implementation.

Alternatively (and perhaps even better), `tf.losses` should sum up the loss for both the real and imaginary components but that seems like a bigger change.

### CODE REVIEW

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -830,6 +831,8 @@ def to_float(x, name="ToFloat"):
   Raises:
     TypeError: If `x` cannot be cast to the `float32`.
   """
+  if x.dtype.is_complex:
+    logging.warn('Casting complex to real discards imaginary part.')
```
Comment (caisq) [inline tensorflow/python/ops/math_ops.py]: The rest of this file uses double quotes. Using single quotes here will cause pylint errors. Please fix.

Comment (carlthome) [reply tensorflow/python/ops/math_ops.py]: Fixed.

**Comment by carlthome** [comment]
Perhaps it would be better to move this warning into tf.cast instead so that `tf.cast(complex_tensor, tf.float32)` will also have a warning and not only `tf.to_float`.

**Comment by ekelsen** [comment]
@caisq Can you update your review - it looks like your comments have been addressed.

**Review by caisq — APPROVED** [review]

**Comment by caisq** [comment]
@carlthome can you fix the merge conflicts?

**Comment by carlthome** [comment]
@caisq fixed merge conflict.

Do you think it would be better to move the warning into `tf.cast` instead of `tf.to_float` so that `tf.cast(complex_tensor, tf.float32)` will also have a warning?

**Comment by caisq** [comment]
@carlthome that sounds good to me. please do the move.

**Comment by carlthome** [comment]
@caisq moved!

**Review by caisq — APPROVED** [review]

**Comment by caisq** [comment]
There are test failures after the move, e.g.,
https://source.cloud.google.com/results/invocations/94c909ec-7af9-45ec-920e-d61d6fd7c838/targets/%2F%2Ftensorflow%2Fpython:adagrad_da_test/log

**Comment by carlthome** [comment]
Sorry, forgot that `tf.cast("hello", tf.int32)` etc. is legal syntax at the moment (!). Moved the warning to after `convert_to_tensor` to let that work.

**Comment by gunan** [comment]
Looks like there are still build issues.
```
  File "/home/kbuilder/.cache/bazel/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/execroot/org_tensorflow/bazel-out/k8-py3-opt/bin/tensorflow/python/kernel_tests/svd_op_test.runfiles/org_tensorflow/tensorflow/python/ops/math_ops.py", line 778, in cast
    if x.dtype.is_complex and dtype.is_floating:
AttributeError: type object 'numpy.complex64' has no attribute 'is_floating'
```

**Comment by carlthome** [comment]
@gunan green now? 

(by the way, can I run the CI test suite locally somehow?)

**Comment by gunan** [comment]
Retriggering tests.
To run tests locally, you can follow the install from sources instructions to set up TF dependencies, and then run `bazel test //tensorflow/python:math_ops_test`

**Pull request merged**